sirepo/auth_role_moderation.py

```diff
@@ -0,0 +1,172 @@
+import sirepo.http_reply
+import sirepo.http_request
+import sirepo.simulation_db
+import sirepo.smtp
+import sirepo.uri
+import sqlalchemy
+
+_cfg = None
+
+@sirepo.api_perm.require_user
+def api_saveModerationReason():
+    def _request_link(sim_type):
+        return sirepo.uri.server_root(external=True).rstrip('/') \
+               + sirepo.uri.local_route(sim_type, 'admRoles')
+
+    def _send_request_email(info):
+        sirepo.smtp.send(
+            recipient=_cfg.moderator_email,
+            subject=f'{info.sim_type} Access Request',
+            body=pkjinja.render_resource('auth_role_moderation/moderation_email', info),
+        )
+
+    req = sirepo.http_request.parse_post(require_sim_type=False)
+    d = req.req_data
+    u = sirepo.auth.logged_in_user()
+    r = sirepo.auth_role.for_sim_type(d.simulationType)
+    with sirepo.util.THREAD_LOCK:
+        if sirepo.auth_db.UserRole.has_role(u, r):
+            raise sirepo.util.Redirect(sirepo.uri.local_route(d.simulationType))
+        try:
+            sirepo.auth_db.UserRoleInvite(
+                uid=u,
+                role=r,
+                status='pending',
+                token=sirepo.util.random_base62(32),
+            ).save()
+        except sqlalchemy.exc.IntegrityError as e:
+            pkdlog('Error={} saving UserRoleInvite for uid={} role={} stack={}', e, u, r, pkdexc())
+            raise sirepo.util.UserAlert(
+                f"You've already submitted a moderation request.",
+            )
+
+    l = _request_link(d.simulationType)
+    if sirepo.auth.dev_and_no_method('email'):
+        pkdlog('Email not enabled. Moderation reason={} Link to approve/deny={}', d.reason, l)
+    else:
+        _send_request_email(
+            PKDict(
+                display_name=sirepo.auth.user_display_name(u),
+                email_addr=sirepo.auth.user_name(),
+                link=l,
+                reason=d.reason,
+                role=sirepo.auth_role.for_sim_type(d.simulationType),
+                sim_type=d.simulationType,
+                uid=u,
+            ).pkupdate(sirepo.http_request.user_agent_headers())
+        )
+    return sirepo.http_reply.gen_json_ok()
+
+
+@sirepo.api_perm.require_adm
+def api_getPendingAndClarifyUserRoleInvites():
+    return sirepo.http_reply.gen_json(PKDict(
+        rows=[r.as_pkdict() for r in (
+                sirepo.auth_db.UserRoleInvite.get_pending_rows()
+                + sirepo.auth_db.UserRoleInvite.get_clarify_rows()
+        )],
+    ))
+
+
+@sirepo.api_perm.require_adm
+def api_approveOrDenyRole():
+    def _send_approve_or_deny_email(info):
+        a = 'Approved' if info.approved else 'Denied'
+        if sirepo.auth.dev_and_no_method('email'):
+            pkdlog('Email not enabled. Access is {}', a)
+            return
+        sirepo.smtp.send(
+            recipient=sirepo.auth.get_module(
+                'email',
+            ).AuthEmailUser.search_by(uid=info.uid).unverified_email,
```

Comment:
This should be user_name, not unverified_email. The user must already be verified before they can get to this point (btw, this should be tested).

Reply:
f912c0cd3b5b5b02720985017a7af542f5452940